### ISSUE TITLE
Feature request: Move ssl certs for dovecot from dovecot.conf to local.conf

### ISSUE THREAD

**Issue by nicutor** [virtualmin/virtualmin-gpl#324]
Hi there,

I am using an automation tool to maintain the dovecot configuration files and I was dealing with this.

1. When a domain is added on virtualmin, the SSLs are added on /etc/dovecot/dovecot.conf:

```
local_name domain.com {
  ssl_cert = </home/domain.com/ssl.combined
  ssl_key = </home/domain.com/ssl.key
}
local_name *.domain.com {
  ssl_cert = </home/domain.com/ssl.combined
  ssl_key = </home/domain.com/ssl.key
}
```

2. The automation tool is re-writing the /etc/dovecot/dovecot.conf and the SSLs are gone.

**Resolution:**

1. Move these settings to /etc/dovecot/local.conf

2. Add inclusion of local.conf on dovecot.conf. Usually this is already there:

```
!include_try local.conf
```

dovecot is the only software left with this issue.

postfix have /etc/postfix/sni_map
proftpd have /etc/proftpd/conf.d/virtualmin.conf
apache have IncludeOptional sites-enabled/*.conf

and so on..

Thank you for your work!

**Comment by nicutor** [comment]
Hi @jcameron, any update on this? I still need to manually move SSLs added by virtualmin to /etc/dovecot/dovecot.conf.
Thank you!

**Comment by jcameron** [comment]
No, it's still in progress..

**Comment by nicutor** [comment]
Still need this for automation. Thank you!

**Comment by iliaross** [comment]
Jamie, do you want me to send a PR for this one?

**Comment by jcameron** [comment]
I don't think any PR is needed, you can just add a line to ` /etc/dovecot/dovecot.conf` like `!include_try virtualmin.conf` . 

And then in `/etc/webmin/dovecot/config` add the line `add_config=/etc/dovecot/virtualmin.conf`

**Comment by iliaross** [comment]
Sadly, the Webmin Dovecot module isn’t designed to work this way. The `dovecot_config` directive expects a complete Dovecot config file to be passed.

So I was thinking of using a separate `/etc/dovecot/conf.d/99-webmin.com` file for that purpose, to create new records.

**Comment by jcameron** [comment]
> Sadly, the Webmin Dovecot module isn’t designed to work this way. The dovecot_config directive expects a complete Dovecot config file to be passed.

Yes, but if the main config references other config files with an `!include` line, they will be follow. So I don't get why suggestion I had above wouldn't work for users who want to put any `local_name` directives in a separate file.

**Comment by nicutor** [comment]
I am so glad that you are working on this guys. Thank you!

**Comment by iliaross** [comment]
> So I don't get why suggestion I had above wouldn't work for users who want to put any `local_name` directives in a separate file.

Yes, it's already automatically included if the file is located in `/etc/dovecot/conf.d` and has a `.conf` extension. Therefore, the options in `/etc/dovecot/conf.d/99-webmin.com.conf` are recognized.

However, in Virtualmin, we do not add those entries to `/etc/dovecot/conf.d/99-webmin.com.conf` or `/etc/dovecot/conf.d/99-virtualmin.com.conf`. The goal is to avoid modifying the standard Dovecot config files that may be updated during package upgrades, and instead, place our custom settings in separate configuration files.

**Comment by jcameron** [comment]
@iliaross are you suggesting making that the default for **new** Virtualmin installs using the `add_config` option? If so, that would be reasonable..

**Comment by iliaross** [comment]
> [@iliaross](https://github.com/iliaross) are you suggesting making that the default for **new** Virtualmin installs using the `add_config` option? If so, that would be reasonable..

I wouldn’t mind setting it as the default for new Virtualmin installs, if that option actually worked.

**Comment by jcameron** [comment]
Oh I see it now .. check out the patch above

**Comment by iliaross** [comment]
Alright, great!

Yet, here are three more patches that finally make everything work as expected. I’ve run quite a few tests which worked just fine for me, and I expect this to be included in Virtualmin 8.2.0 for all new installs.

For the record, users shouldn’t manually configure `add_config` for existing installs unless they also move all existing `local_name` directives from `dovecot.conf` to `conf.d/99-virtualmin.conf`. Otherwise, it might break the order of those entries. We should all keep in mind that the `!include` for this custom `conf.d/99-virtualmin.conf` config will be processed before all the `local_name` entries in `dovecot.conf`.

Jamie, could you please double-check this https://github.com/virtualmin/virtualmin-gpl/commit/04e683f9c264910316b8e4d5a9cde5c63cd165ed patch in Virtualmin? I’m 99% sure it’s correct, but I want to be certain.

**Comment by jcameron** [comment]
> Jamie, could you please double-check this https://github.com/virtualmin/virtualmin-gpl/commit/04e683f9c264910316b8e4d5a9cde5c63cd165ed patch in Virtualmin? I’m 99% sure it’s correct, but I want to be certain.

Try running `virtualmin functional-test --test sslserv` with this change to validate if it works

**Comment by iliaross** [comment]
> Try running `virtualmin functional-test --test sslserv` with this change to validate if it works

Yeah, it did work for this specific change, e.g.:

```
root@ubuntu26-pro:/etc# virtualmin functional-test --test sslserv
Running sslserv tests ..
    Running create-domain.pl --domain example.com --desc 'Test SSL domain' --pass smeg --dir --unix --web --dns --mail --ssl --logrotate --webmin --allocate-ip --content 'Test SSL home page' --limits-from-plan --no-email --no-slaves --no-secondaries --default-cert-owner ..
    .. success
    Running create-domain.pl --domain example.net --alias example.com --prefix example2 --desc 'Test alias-domain' --dir --web --dns --mail --limits-from-plan --no-email --no-slaves --no-secondaries --default-cert-owner ..
    .. success
    Running install-cert.pl --domain example.com --ca /usr/share/webmin/virtual-server/lets-encrypt-*.pem.txt ..
    .. success
    Running list-domains.pl --ip-only --domain example.com ..
    .. saved PRIVATE_IP value 10.211.55.58
    .. success
    Running install-service-cert.pl --domain example.com --add-domain --service webmin --service usermin --service dovecot --service postfix ..
    .. success
    Running list-domains.pl --multiline --domain example.com ..
    .. success
    Running cat /etc/dovecot/dovecot.conf /etc/dovecot/conf.d/99-virtualmin.conf ..
    .. success
    Running cd / ; wget -O - --cache=off --proxy=off --no-check-certificate --user-agent=Webmin --no-check-certificate --user example --password smeg https://example.com:10000/ ..
    .. success
    Running openssl s_client -host $PRIVATE_IP -servername example.com -port 10000 </dev/null ..
    .. success
    Running openssl s_client -host $PRIVATE_IP -servername example.com -port 20000 </dev/null ..
    .. success
```

**Comment by jcameron** [comment]
I tested it too, and it works fine .. thanks for the patch!

**Comment by iliaross** [comment]
Excellent! Good to know. Thanks!
